From ce7997b4c3e916d191a59d70bb75c3d085e20f94 Mon Sep 17 00:00:00 2001
From: Wendy Liang <wendy.liang@xilinx.com>
Date: Wed, 16 Sep 2020 16:56:20 -0700
Subject: [PATCH 1606/1852] misc: xilinx-ai-engine: clear cached events of
 partition in reset/release

commit 9d4883a20b9041ebd38bb3e8c5c4f79478c201c2 from
https://github.com/Xilinx/linux-xlnx.git

Clear the cached events of an AI engine partition in partition reset
and partition file release function. Otherwise, next time, when the
partition is requested again or after reset, the application will still
see the errors before the last reset/release.

Signed-off-by: Wendy Liang <wendy.liang@xilinx.com>
Reviewed-by: Hyun Kwon <hyun.kwon@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 drivers/misc/xilinx-ai-engine/ai-engine-internal.h  |  1 +
 drivers/misc/xilinx-ai-engine/ai-engine-interrupt.c | 13 +++++++++++++
 drivers/misc/xilinx-ai-engine/ai-engine-part.c      |  3 +++
 drivers/misc/xilinx-ai-engine/ai-engine-reset.c     |  2 ++
 4 files changed, 19 insertions(+)

diff --git a/drivers/misc/xilinx-ai-engine/ai-engine-internal.h b/drivers/misc/xilinx-ai-engine/ai-engine-internal.h
index b119a94ee56e..ac2f3e63ed55 100644
--- a/drivers/misc/xilinx-ai-engine/ai-engine-internal.h
+++ b/drivers/misc/xilinx-ai-engine/ai-engine-internal.h
@@ -546,6 +546,7 @@ int aie_device_init(struct aie_device *adev);
 
 void aie_array_backtrack(struct work_struct *work);
 irqreturn_t aie_interrupt(int irq, void *data);
+void aie_part_clear_cached_events(struct aie_partition *apart);
 
 bool aie_part_has_mem_mmapped(struct aie_partition *apart);
 bool aie_part_has_regs_mmapped(struct aie_partition *apart);
diff --git a/drivers/misc/xilinx-ai-engine/ai-engine-interrupt.c b/drivers/misc/xilinx-ai-engine/ai-engine-interrupt.c
index f8d2f71df73a..77985118f014 100644
--- a/drivers/misc/xilinx-ai-engine/ai-engine-interrupt.c
+++ b/drivers/misc/xilinx-ai-engine/ai-engine-interrupt.c
@@ -825,6 +825,19 @@ static u32 aie_get_module_errors(struct aie_partition *apart,
 	return num_err;
 }
 
+/**
+ * aie_part_clear_cached_events() - clear cached events in a partition.
+ * @apart: AIE partition pointer
+ *
+ * This function will clear the cached events in a partition.
+ */
+void aie_part_clear_cached_events(struct aie_partition *apart)
+{
+	aie_resource_clear_all(&apart->core_event_status);
+	aie_resource_clear_all(&apart->mem_event_status);
+	aie_resource_clear_all(&apart->pl_event_status);
+}
+
 /**
  * aie_register_error_notification() - register a callback for error
  *				       notification.
diff --git a/drivers/misc/xilinx-ai-engine/ai-engine-part.c b/drivers/misc/xilinx-ai-engine/ai-engine-part.c
index 39c6173c687d..a4a55637084f 100644
--- a/drivers/misc/xilinx-ai-engine/ai-engine-part.c
+++ b/drivers/misc/xilinx-ai-engine/ai-engine-part.c
@@ -335,6 +335,9 @@ static int aie_part_release(struct inode *inode, struct file *filp)
 	apart->error_cb.cb = NULL;
 	apart->error_cb.priv = NULL;
 	apart->status = 0;
+
+	aie_part_clear_cached_events(apart);
+
 	mutex_unlock(&apart->mlock);
 
 	return 0;
diff --git a/drivers/misc/xilinx-ai-engine/ai-engine-reset.c b/drivers/misc/xilinx-ai-engine/ai-engine-reset.c
index 63ec9bc095d4..a0aab0b538dc 100644
--- a/drivers/misc/xilinx-ai-engine/ai-engine-reset.c
+++ b/drivers/misc/xilinx-ai-engine/ai-engine-reset.c
@@ -208,6 +208,8 @@ int aie_part_reset(struct aie_partition *apart)
 
 	aie_part_set_cols_clkbuf(apart, false);
 
+	aie_part_clear_cached_events(apart);
+
 	mutex_unlock(&apart->mlock);
 	return 0;
 }
-- 
2.31.1

